File: MyChores/ViewModels/HouseholdViewModel.swift
// HouseholdViewModel.swift
// MyChores
//
// Created on 2025-05-02.
//

import Foundation
import Combine
import FirebaseFirestore

/// ViewModel for household-related views
class HouseholdViewModel: ObservableObject {
    // MARK: - Published Properties
    
    /// List of households
    @Published var households: [Household] = []
    
    /// Currently selected household
    @Published var selectedHousehold: Household?
    
    /// Members of the current household
    @Published var householdMembers: [User] = []
    
    /// Current logged-in user
    @Published var currentUser: User?

    /// Rules for the selected household (NEW)
    @Published var householdRules: [HouseholdRule] = []
    
    /// Loading state
    @Published var isLoading = false
    
    /// Error message
    @Published var errorMessage: String?
    
    // MARK: - Private Properties
    
    /// Household service instance
    private let householdService: HouseholdServiceProtocol
    
    /// User service instance
    private let userService: UserServiceProtocol

    /// Auth service instance
    private let authService: any AuthServiceProtocol
    
    // MARK: - Initialization
    
    init(selectedHouseholdId: String? = nil, householdService: HouseholdServiceProtocol = HouseholdService.shared, userService: UserServiceProtocol = UserService.shared, authService: any AuthServiceProtocol = AuthService.shared) {
        self.householdService = householdService
        self.userService = userService
        self.authService = authService
        
        // Set current user from authService
        if let firebaseUser = authService.currentUser {
            self.currentUser = firebaseUser // Assuming authService.currentUser is your app's User model
        }
        
        if let initialId = selectedHouseholdId {
            fetchHousehold(id: initialId)
        } else {
            // If no specific household is pre-selected, try to load the user's households
            // and potentially select the first one or a previously selected one.
            loadUserHouseholdsAndSelectDefault()
        }
    }
    
    private func loadUserHouseholdsAndSelectDefault() {
        guard let userId = authService.getCurrentUserId() else {
            errorMessage = "User not authenticated."
            return
        }
        isLoading = true
        Task {
            do {
                let fetchedHouseholds = try await householdService.fetchHouseholds(forUserId: userId)
                await MainActor.run {
                    self.households = fetchedHouseholds
                    if let firstHousehold = fetchedHouseholds.first {
                        self.selectHousehold(firstHousehold) // Automatically select the first household
                    } else {
                        // Handle case where user has no households yet
                        self.selectedHousehold = nil
                        self.householdMembers = []
                        self.householdRules = [] // Clear rules if no household
                    }
                    self.isLoading = false
                }
            } catch {
                await MainActor.run {
                    self.errorMessage = "Failed to load households: \(error.localizedDescription)"
                    self.isLoading = false
                }
            }
        }
    }

    /// Selects a household and fetches its details, members, and rules.
    func selectHousehold(_ household: Household) {
        self.selectedHousehold = household
        if let householdId = household.id {
            fetchHouseholdDetails(id: householdId)
            loadHouseholdRules() // Load rules when a household is selected
        }
    }
    
    // MARK: - Household Methods
    
    /// Load all households for the current user
    func loadHouseholds() {
        guard let userId = authService.getCurrentUserId() else {
            errorMessage = "User not logged in."
            isLoading = false
            return
        }
        
        isLoading = true
        errorMessage = nil
        
        Task {
            do {
                let households = try await householdService.fetchHouseholds(forUserId: userId)
                await MainActor.run {
                    self.households = households
                    if selectedHousehold == nil, let firstHousehold = households.first {
                        self.selectedHousehold = firstHousehold
                        if let currentSelectedHousehold = self.selectedHousehold {
                             self.loadHouseholdMembers(household: currentSelectedHousehold)
                        }
                    }
                    self.isLoading = false
                }
            } catch {
                await MainActor.run {
                    self.errorMessage = "Failed to load households: \(error.localizedDescription)"
                    self.isLoading = false
                }
            }
        }
    }
    
    /// Load a specific household
    /// - Parameters:
    ///   - id: Household ID
    ///   - completion: Optional completion handler with success boolean
    func fetchHousehold(id: String, completion: ((Bool) -> Void)? = nil) {
        isLoading = true
        Task {
            do {
                let fetchedHousehold = try await householdService.fetchHousehold(withId: id)
                await MainActor.run {
                    if let household = fetchedHousehold {
                        self.selectHousehold(household) // Use selectHousehold to also load members and rules
                        completion?(true)
                    } else {
                        self.errorMessage = "Household with ID \(id) not found."
                        completion?(false)
                    }
                    isLoading = false
                }
            } catch {
                await MainActor.run {
                    self.errorMessage = "Failed to fetch household: \(error.localizedDescription)"
                    isLoading = false
                    completion?(false)
                }
            }
        }
    }

    /// Fetches details for the currently selected household, including its members.
    private func fetchHouseholdDetails(id: String) {
        // This function is now primarily for fetching members, as household object is already set.
        // Rules are fetched by loadHouseholdRules()
        isLoading = true
        Task {
            do {
                let members = try await userService.fetchUsers(inHousehold: id)
                await MainActor.run {
                    self.householdMembers = members
                    isLoading = false
                }
            } catch {
                await MainActor.run {
                    self.errorMessage = "Failed to fetch household members: \(error.localizedDescription)"
                    isLoading = false
                }
            }
        }
    }
    
    /// Create a new household
    /// - Parameters:
    ///   - name: Household name
    ///   - completion: Optional completion handler with success boolean
    func createHousehold(name: String, completion: ((Bool) -> Void)? = nil) {
        guard let userId = authService.getCurrentUserId() else {
            errorMessage = "User not logged in."
            isLoading = false
            completion?(false)
            return
        }
        
        isLoading = true
        errorMessage = nil
        
        Task {
            do {
                let newHousehold = try await householdService.createHousehold(name: name, ownerUserId: userId)
                await MainActor.run {
                    self.households.append(newHousehold)
                    self.selectedHousehold = newHousehold // Optionally select the new household
                    if let currentSelectedHousehold = self.selectedHousehold {
                        self.loadHouseholdMembers(household: currentSelectedHousehold)
                    }
                    self.isLoading = false
                    completion?(true)
                }
            } catch {
                await MainActor.run {
                    self.errorMessage = "Failed to create household: \(error.localizedDescription)"
                    self.isLoading = false
                    completion?(false)
                }
            }
        }
    }
    
    /// Join a household using an invite code
    /// - Parameters:
    ///   - inviteCode: Household invite code
    ///   - completion: Optional completion handler with success boolean
    func joinHousehold(inviteCode: String, completion: ((Bool) -> Void)? = nil) {
        guard let userId = authService.getCurrentUserId() else {
            errorMessage = "User not logged in."
            isLoading = false
            completion?(false)
            return
        }
        
        isLoading = true
        errorMessage = nil
        
        Task {
            do {
                guard let householdToJoin = try await householdService.findHousehold(byInviteCode: inviteCode) else {
                    await MainActor.run {
                        self.errorMessage = "Invalid invite code."
                        self.isLoading = false
                        completion?(false)
                    }
                    return
                }
                try await householdService.addMember(userId: userId, toHouseholdId: householdToJoin.id ?? "")
                try await userService.addUserToHousehold(userId: userId, householdId: householdToJoin.id ?? "")
                
                await MainActor.run {
                    // Refresh households list or add the joined household
                    self.loadHouseholds() 
                    self.isLoading = false
                    completion?(true)
                }
            } catch {
                await MainActor.run {
                    self.errorMessage = "Failed to join household: \(error.localizedDescription)"
                    self.isLoading = false
                    completion?(false)
                }
            }
        }
    }
    
    /// Update a household's name
    /// - Parameters:
    ///   - householdId: Household ID
    ///   - newName: New household name
    func updateHouseholdName(householdId: String, newName: String) {
        isLoading = true
        errorMessage = nil
        
        Task {
            do {
                try await householdService.updateHouseholdName(householdId: householdId, newName: newName)
                await MainActor.run {
                    // Refresh the specific household or the list
                    if let index = self.households.firstIndex(where: { $0.id == householdId }) {
                        self.households[index].name = newName
                    }
                    if self.selectedHousehold?.id == householdId {
                        self.selectedHousehold?.name = newName
                    }
                    self.isLoading = false
                }
            } catch {
                await MainActor.run {
                    self.errorMessage = "Failed to update household name: \(error.localizedDescription)"
                    self.isLoading = false
                }
            }
        }
    }
    
    /// Leave a household
    /// - Parameters:
    ///   - householdId: Household ID
    ///   - completion: Optional completion handler with success boolean
    func leaveHousehold(householdId: String, completion: ((Bool) -> Void)? = nil) {
        guard let userId = authService.getCurrentUserId() else {
            errorMessage = "User not authenticated."
            completion?(false)
            return
        }
        isLoading = true
        Task {
            do {
                try await householdService.removeMember(userId: userId, fromHouseholdId: householdId)
                await MainActor.run {
                    // Refresh user's households list and potentially select another one or show empty state
                    self.loadUserHouseholdsAndSelectDefault()
                    isLoading = false
                    completion?(true)
                }
            } catch {
                await MainActor.run {
                    self.errorMessage = "Failed to leave household: \(error.localizedDescription)"
                    isLoading = false
                    completion?(false)
                }
            }
        }
    }
    
    // MARK: - Household Rules Methods (NEW)

    /// Load rules for the currently selected household
    func loadHouseholdRules() {
        guard let householdId = selectedHousehold?.id else {
            self.householdRules = [] // Clear rules if no household is selected
            // errorMessage = "No household selected to load rules from."
            return
        }
        isLoading = true
        Task {
            do {
                let rules = try await householdService.fetchHouseholdRules(forHouseholdId: householdId)
                await MainActor.run {
                    self.householdRules = rules.sorted(by: { ($0.displayOrder ?? Int.max) < ($1.displayOrder ?? Int.max) }) // Sort by displayOrder, then createdAt if displayOrder is nil or same
                    // As a secondary sort, if displayOrder is not used or items have same displayOrder:
                    // self.householdRules = rules.sorted(by: { $0.createdAt < $1.createdAt })
                    isLoading = false
                }
            } catch {
                await MainActor.run {
                    self.errorMessage = "Failed to load household rules: \(error.localizedDescription)"
                    isLoading = false
                }
            }
        }
    }

    /// Add a new rule to the selected household
    func addHouseholdRule(ruleText: String) {
        guard let householdId = selectedHousehold?.id, let userId = authService.getCurrentUserId() else {
            errorMessage = "Cannot add rule: No household selected or user not authenticated."
            return
        }
        guard !ruleText.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            errorMessage = "Rule text cannot be empty."
            return
        }
        
        isLoading = true
        Task {
            do {
                let newRule = try await householdService.createHouseholdRule(householdId: householdId, ruleText: ruleText, createdByUserId: userId)
                await MainActor.run {
                    self.householdRules.append(newRule)
                    // Optionally re-sort or just append
                    self.householdRules.sort(by: { ($0.displayOrder ?? Int.max) < ($1.displayOrder ?? Int.max) })
                    isLoading = false
                }
            } catch {
                await MainActor.run {
                    self.errorMessage = "Failed to add household rule: \(error.localizedDescription)"
                    isLoading = false
                }
            }
        }
    }

    /// Update an existing household rule
    func updateHouseholdRule(rule: HouseholdRule, newText: String? = nil, newDisplayOrder: Int? = nil) {
        guard let householdId = selectedHousehold?.id else {
            errorMessage = "Cannot update rule: No household selected."
            return
        }
        
        var ruleToUpdate = rule
        if let text = newText, !text.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            ruleToUpdate.ruleText = text
        }
        if let order = newDisplayOrder {
            ruleToUpdate.displayOrder = order
        }

        isLoading = true
        Task {
            do {
                let updatedRule = try await householdService.updateHouseholdRule(ruleToUpdate)
                await MainActor.run {
                    if let index = self.householdRules.firstIndex(where: { $0.id == updatedRule.id }) {
                        self.householdRules[index] = updatedRule
                        self.householdRules.sort(by: { ($0.displayOrder ?? Int.max) < ($1.displayOrder ?? Int.max) })
                    }
                    isLoading = false
                }
            } catch {
                await MainActor.run {
                    self.errorMessage = "Failed to update household rule: \(error.localizedDescription)"
                    isLoading = false
                }
            }
        }
    }

    /// Delete a household rule
    func deleteHouseholdRule(rule: HouseholdRule) {
        guard let ruleId = rule.id, let householdId = selectedHousehold?.id else {
            errorMessage = "Cannot delete rule: Rule ID or Household ID missing."
            return
        }
        isLoading = true
        Task {
            do {
                // Using the corrected service method that requires householdId
                try await householdService.deleteHouseholdRule(ruleId: ruleId)
                await MainActor.run {
                    self.householdRules.removeAll { $0.id == ruleId }
                    isLoading = false
                }
            } catch {
                await MainActor.run {
                    self.errorMessage = "Failed to delete household rule: \(error.localizedDescription)"
                    isLoading = false
                }
            }
        }
    }
}

File: MyChoresAnd/app/src/main/java/com/example/mychoresand/viewmodels/HouseholdViewModel.kt
package com.example.mychoresand.viewmodels

import android.util.Log
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.example.mychoresand.di.AppContainer
import com.example.mychoresand.models.Household
import com.example.mychoresand.models.User
import com.example.mychoresand.services.AuthService
import com.example.mychoresand.services.HouseholdService
import com.example.mychoresand.services.UserService
import com.example.mychoresand.utils.PreferencesManager
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch

/**
 * ViewModel for household-related views
 */
class HouseholdViewModel(
    private val householdService: HouseholdService,
    private val userService: UserService,
    private val authService: AuthService
) : ViewModel() {
    
    companion object {
        private const val TAG = "HouseholdViewModel"
    }
    
    // StateFlow for households
    private val _households = MutableStateFlow<List<Household>>(emptyList())
    val households: StateFlow<List<Household>> = _households
    
    // StateFlow for selected household
    private val _selectedHousehold = MutableStateFlow<Household?>(null)
    val selectedHousehold: StateFlow<Household?> = _selectedHousehold
    
    // StateFlow for household members - match iOS property name
    private val _householdMembers = MutableStateFlow<List<User>>(emptyList())
    val householdMembers: StateFlow<List<User>> = _householdMembers
    
    // StateFlow for current user - match iOS property name
    private val _currentUser = MutableStateFlow<User?>(null)
    val currentUser: StateFlow<User?> = _currentUser
    
    // StateFlow for loading state
    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading
    
    // StateFlow for error message
    private val _errorMessage = MutableStateFlow<String?>(null)
    val errorMessage: StateFlow<String?> = _errorMessage
    
    init {
        loadCurrentUser()
        loadHouseholds()
    }
    
    /**
     * Load the current user
     */
    private fun loadCurrentUser() {
        viewModelScope.launch {
            val uid = authService.currentUser?.uid ?: return@launch
            
            try {
                // Use the flow-based API
                userService.getUser(uid).collect { user ->
                    _currentUser.value = user
                }
            } catch (e: Exception) {
                Log.e(TAG, "Error loading current user: ${e.message}", e)
            }
        }
    }
    
    /**
     * Load all households for the current user
     */
    fun loadHouseholds() {
        viewModelScope.launch {
            _isLoading.value = true
            _errorMessage.value = null
            
            try {
                if (authService.currentUser == null) {
                    throw IllegalStateException("Not signed in")
                }
                
                // Use the flow-based API
                householdService.getUserHouseholds().collect { fetchedHouseholds ->
                    _households.value = fetchedHouseholds
                    
                    // If we only have one household, select it automatically
                    if (fetchedHouseholds.size == 1) {
                        _selectedHousehold.value = fetchedHouseholds.first()
                        fetchedHouseholds.first().id?.let { householdId ->
                            loadHouseholdMembers(householdId)
                            // Save current household ID to preferences
                            AppContainer.preferencesManager.saveString(
                                PreferencesManager.KEY_CURRENT_HOUSEHOLD_ID, 
                                householdId
                            )
                        }
                    }
                    
                    _isLoading.value = false
                }
            } catch (e: Exception) {
                Log.e(TAG, "Error loading households: ${e.message}", e)
                _errorMessage.value = "Failed to load households: ${e.message}"
                _isLoading.value = false
            }
        }
    }
    
    /**
     * Load a specific household
     */
    fun fetchHousehold(householdId: String) {
        viewModelScope.launch {
            _isLoading.value = true
            _errorMessage.value = null
            
            try {
                // Use the flow-based API
                householdService.getHousehold(householdId).collect { household ->
                    household?.let {
                        _selectedHousehold.value = it
                        loadHouseholdMembers(householdId)
                    } ?: run {
                        _errorMessage.value = "Household not found"
                    }
                    
                    _isLoading.value = false
                }
            } catch (e: Exception) {
                Log.e(TAG, "Error loading household: ${e.message}", e)
                _errorMessage.value = "Failed to load household: ${e.message}"
                _isLoading.value = false
            }
        }
    }
    
    /**
     * Clear any error message
     */
    fun clearErrorMessage() {
        _errorMessage.value = null
    }
    
    /**
     * Create a new household
     */
    fun createHousehold(name: String, description: String = "", callback: (Result<String?>) -> Unit = {}) {
        viewModelScope.launch {
            _isLoading.value = true
            _errorMessage.value = null
            
            try {
                val result = householdService.createHousehold(name, description)
                
                result.fold(
                    onSuccess = { household ->
                        // Add to list and select it
                        val updatedHouseholds = _households.value.toMutableList()
                        updatedHouseholds.add(household)
                        _households.value = updatedHouseholds
                        _selectedHousehold.value = household
                        
                        // Save current household ID to preferences
                        household.id?.let { householdId ->
                            AppContainer.preferencesManager.saveString(
                                PreferencesManager.KEY_CURRENT_HOUSEHOLD_ID, 
                                householdId
                            )
                            // Load members for the new household
                            loadHouseholdMembers(householdId)
                        }
                        
                        // Call the callback with success
                        callback(Result.success(household.id))
                    },
                    onFailure = { e ->
                        callback(Result.failure(e))
                        throw e
                    }
                )
            } catch (e: Exception) {
                Log.e(TAG, "Error creating household: ${e.message}", e)
                _errorMessage.value = "Failed to create household: ${e.message}"
                callback(Result.failure(e))
            } finally {
                _isLoading.value = false
            }
        }
    }
    
    /**
     * Join a household using an invite code
     */
    fun joinHousehold(inviteCode: String, callback: (Result<String?>) -> Unit = {}) {
        viewModelScope.launch {
            _isLoading.value = true
            _errorMessage.value = null
            
            try {
                val result = householdService.joinHousehold(inviteCode)
                
                result.fold(
                    onSuccess = { household ->
                        // Reload households to get updated list
                        loadHouseholds()
                        // Select the joined household
                        _selectedHousehold.value = household
                        
                        // Save current household ID to preferences
                        household.id?.let { householdId ->
                            AppContainer.preferencesManager.saveString(
                                PreferencesManager.KEY_CURRENT_HOUSEHOLD_ID, 
                                householdId
                            )
                            // Load members for the joined household
                            loadHouseholdMembers(householdId)
                        }
                        
                        // Call the callback with success
                        callback(Result.success(household.id))
                    },
                    onFailure = { e ->
                        callback(Result.failure(e))
                        throw e
                    }
                )
            } catch (e: Exception) {
                Log.e(TAG, "Error joining household: ${e.message}", e)
                _errorMessage.value = "Failed to join household: ${e.message}"
                _isLoading.value = false
                callback(Result.failure(e))
            }
        }
    }
    
    /**
     * Leave a household
     */
    fun leaveHousehold(householdId: String, onComplete: ((hasHouseholds: Boolean) -> Unit)? = null) {
        viewModelScope.launch {
            _isLoading.value = true
            _errorMessage.value = null
            
            try {
                val result = householdService.leaveHousehold(householdId)
                
                result.fold(
                    onSuccess = {
                        // Clear selected household if it was the one we left
                        if (_selectedHousehold.value?.id == householdId) {
                            _selectedHousehold.value = null
                            // Clear current household ID from preferences
                            AppContainer.preferencesManager.remove(PreferencesManager.KEY_CURRENT_HOUSEHOLD_ID)
                        }
                        
                        // Reload households to get updated list and check if any remain
                        householdService.getUserHouseholds().collect { updatedHouseholds ->
                            _households.value = updatedHouseholds
                            // Notify caller if user has any households left
                            onComplete?.invoke(updatedHouseholds.isNotEmpty())
                            _isLoading.value = false
                        }
                    },
                    onFailure = { e ->
                        _isLoading.value = false
                        onComplete?.invoke(_households.value.isNotEmpty())
                        throw e
                    }
                )
            } catch (e: Exception) {
                Log.e(TAG, "Error leaving household: ${e.message}", e)
                _errorMessage.value = "Failed to leave household: ${e.message}"
                _isLoading.value = false
                // Call callback with current household state in case of error
                onComplete?.invoke(_households.value.isNotEmpty())
            }
        }
    }
    
    /**
     * Load members of a household
     */
    private fun loadHouseholdMembers(householdId: String) {
        viewModelScope.launch {
            try {
                // Use the flow-based API
                userService.getHouseholdUsers(householdId).collect { members ->
                    _householdMembers.value = members
                }
            } catch (e: Exception) {
                Log.e(TAG, "Error loading household members: ${e.message}", e)
                _errorMessage.value = "Failed to load members: ${e.message}"
            }
        }
    }
    
    /**
     * Check if the current user is the owner of a household
     */
    fun isCurrentUserOwner(household: Household): Boolean {
        val userId = authService.currentUser?.uid ?: return false
        return household.ownerUserId == userId
    }
    
    /**
     * Select a household and load its data
     * @param household The household to select
     */
    fun selectHousehold(household: Household) {
        viewModelScope.launch {
            _selectedHousehold.value = household
            
            household.id?.let { householdId ->
                // Save current household ID to preferences
                AppContainer.preferencesManager.saveString(
                    PreferencesManager.KEY_CURRENT_HOUSEHOLD_ID,
                    householdId
                )
                
                // Load household members
                loadHouseholdMembers(householdId)
                
                // Load chores for the selected household using ChoreViewModel
                AppContainer.choreViewModel.loadHouseholdChores(householdId)
            }
        }
    }
}
